### PR TITLE
Exposes CA Certificate configuration value

### PR DESCRIPTION
**Pull request by kunickiaj**
Adds the ability to set a CA certificate for verifying the connection to Vault.

### CODE REVIEW

File: src/main/java/com/datapipe/jenkins/vault/configuration/VaultConfiguration.java [hunk 1]
```diff
@@ -70,6 +74,8 @@ public String getVaultUrl() {
         return vaultUrl;
     }
 
+    public String getVaultCACert() { return vaultCACert; }
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/configuration/VaultConfiguration.java]: Should be a secret
```suggestion
    public Secret getVaultCACert() { return vaultCACert; }
```

File: src/main/java/com/datapipe/jenkins/vault/configuration/VaultConfiguration.java [hunk 2]
```diff
@@ -79,6 +85,9 @@ public void setVaultUrl(String vaultUrl) {
         this.vaultUrl = normalizeUrl(fixEmptyAndTrim(vaultUrl));
     }
 
+    @DataBoundSetter
+    public void setVaultCACert(String vaultCACert) { this.vaultCACert = vaultCACert; }
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/configuration/VaultConfiguration.java]: Should be a secret
```suggestion
    public void setVaultCACert(Secret vaultCACert) { this.vaultCACert = vaultCACert; }
```

File: src/main/resources/com/datapipe/jenkins/vault/configuration/VaultConfiguration/config.jelly [hunk 3]
```diff
@@ -4,7 +4,10 @@
     <f:entry title="Vault URL">
         <f:textbox field="vaultUrl" name="vaultUrl"/>
     </f:entry>
-    <f:entry title="Vault Credential" field="vaultCredentialId">
+    <f:entry title="Vault CA Certificate">
+        <f:textarea field="vaultCACert" name="vaultCACert"/>
+    </f:entry>
+    <f:entry title="Vault Credential" field="vaultCredentialId" name="vaultCredentialId">
```
Comment (casz) [inline src/main/resources/com/datapipe/jenkins/vault/configuration/VaultConfiguration/config.jelly]: Please revert 😅 Looks like a failed merge

File: src/main/resources/com/datapipe/jenkins/vault/configuration/VaultConfiguration/config.jelly [hunk 4]
```diff
@@ -4,7 +4,10 @@
     <f:entry title="Vault URL">
         <f:textbox field="vaultUrl" name="vaultUrl"/>
     </f:entry>
-    <f:entry title="Vault Credential" field="vaultCredentialId">
+    <f:entry title="Vault CA Certificate">
+        <f:textarea field="vaultCACert" name="vaultCACert"/>
```
Comment (casz) [inline src/main/resources/com/datapipe/jenkins/vault/configuration/VaultConfiguration/config.jelly]: I suggest putting it under advanced.
Also best depend on multiline-secrets-ui
```
<dependency>
  <!-- TODO: after upgrading jenkins.version >= 2.171, migrate dependency -->
  <groupId>io.jenkins.temp.jelly</groupId>
  <artifactId>multiline-secrets-ui</artifactId>
  <version>1.0</version>
</dependency>
```

It should look like this:
```xml
<f:entry title="Vault CA Certificate" field="vaultCACert">
  <secretTextarea xmlns="/io/jenkins/temp/jelly"/>
</f:entry>
```

File: src/test/java/com/datapipe/jenkins/vault/it/folder/FolderIT.java [hunk 5]
```diff
@@ -46,11 +36,19 @@
 import java.util.List;
 import java.util.Map;
 
-import hudson.model.FreeStyleBuild;
-import hudson.model.FreeStyleProject;
-import hudson.model.Result;
-import hudson.tasks.Shell;
-import jenkins.model.GlobalConfiguration;
+import static com.datapipe.jenkins.vault.it.VaultConfigurationIT.GLOBAL_CREDENTIALS_ID_1;
```
Comment (casz) [inline src/test/java/com/datapipe/jenkins/vault/it/folder/FolderIT.java]: Please don't mess with import order.

I will have a PR up shortly that fixes import/code style.

File: src/test/java/com/datapipe/jenkins/vault/it/VaultConfigurationIT.java [hunk 6]
```diff
@@ -43,12 +37,16 @@
 import java.util.List;
 import java.util.Map;
 
-import hudson.model.FreeStyleBuild;
-import hudson.model.FreeStyleProject;
-import hudson.model.Result;
-import hudson.tasks.Shell;
-import hudson.util.Secret;
-import jenkins.model.GlobalConfiguration;
+import static org.hamcrest.MatcherAssert.assertThat;
```
Comment (casz) [inline src/test/java/com/datapipe/jenkins/vault/it/VaultConfigurationIT.java]: Please don't mess with import order.

I will have a PR up shortly that fixes import/code style.

File: src/main/java/com/datapipe/jenkins/vault/VaultBuildWrapper.java [hunk 7]
```diff
@@ -123,14 +123,15 @@ public void setVaultAccessor(VaultAccessor vaultAccessor) {
 
     private List<LogicalResponse> provideEnvironmentVariablesFromVault(Context context, Run build) {
         String url = getConfiguration().getVaultUrl();
+        final String vaultCACert = getConfiguration().getVaultCACert();
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/VaultBuildWrapper.java]:
```suggestion
        final String vaultCACert = getConfiguration().getVaultCACert().getPlainText();
```

**Comment by kunickiaj** [comment]
Ah sorry, damn IDE settings lemme take care of these.

Happy to change the CA cert field to a secret, though it’s not particularly
sensitive as a public key.

**Comment by casz** [comment]
D0h public key :) let me update my suggestions

**Comment by kunickiaj** [comment]
Ok, updated the patch. Hopefully this looks a bit better.

File: src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java [hunk 8]
```diff
@@ -20,22 +20,26 @@
...
     public void init(String url, VaultCredential credential) {
-        init(url, credential, false);
+        init(url, credential, false, null);
     }
 
     public void init(String url, VaultCredential credential, boolean skipSslVerification) {
+        init(url, credential, skipSslVerification, null);
+    }
+
+    public void init(String url, VaultCredential credential, boolean skipSslVerification, String vaultCACert) {
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java]: Can we switch to a builder pattern this is getting crazy 😆 
I know #34 is also mocking about there
Actually, I like #34 approach more using a VaultConfig

File: src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java [hunk 9]
```diff
@@ -20,22 +20,26 @@
...
-        init(url, null, false);
+        init(url, null, false, null);
     }
 
     public void init(String url, boolean skipSslVerification) {
-        init(url, null, skipSslVerification);
+        init(url, null, skipSslVerification, null);
     }
 
     public void init(String url, VaultCredential credential) {
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java]: deprecate and switch to builder

File: src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java [hunk 10]
```diff
@@ -20,22 +20,26 @@
 
 
     public void init(String url) {
-        init(url, null, false);
+        init(url, null, false, null);
     }
 
     public void init(String url, boolean skipSslVerification) {
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java]: deprecate and switch to builder

File: src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java [hunk 11]
```diff
@@ -20,22 +20,26 @@
 
 
     public void init(String url) {
```
Comment (casz) [inline src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java]: deprecate and switch to builder

File: src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java (hunk 8 above)
Comment (kunickiaj) [reply src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java]: So, do you want to merge #34 first? Would prefer not have to keep refactoring things.

Comment (casz) [reply src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java]: #34 is not ready for merge, I asked for changes :swe

Comment (casz) [reply src/main/java/com/datapipe/jenkins/vault/VaultAccessor.java]: Also the PR have conflicts, thanks to me trying to fixup the project, so sorry 😓

**Comment by casz** [comment]
@kunickiaj would appreciate if you could enable allow maintainer commit access 😅

**Comment by casz** [comment]
i think #57 replaces both #53 and this PR

**Comment by casz** [comment]
https://github.com/jenkinsci/hashicorp-vault-plugin/releases/tag/hashicorp-vault-plugin-3.0.0 should allow you to load the certificate into the default truststore

**Pull request closed without merging**